File: kadanes/MaxSumCircular.scala
object MaxSumCircular {

    def maxSumCirc(lst: Array[Int]): Int = {

        var globalMax = lst(0)
        var currMax = -9999999
        var currMin = 9999999
        var globalMin = lst(0)
        var total = 0

        for (i <- lst.indices) {
            currMin = math.min(currMin + lst(i), lst(i))
            currMax = math.max(currMax + lst(i), lst(i))

            total += lst(i)

            globalMax = math.max(globalMax, currMax)
            globalMin = math.min(globalMin, currMin)

        }
        if (globalMax > 0) math.max(globalMax, total - globalMin) else globalMax
    }

    def main(args: Array[String]): Unit = {
        val lst = Array(-2,4,-5,4,-5,9,4)
        println(maxSumCirc(lst))
    }

}

File: kadanes/max_sum_circular.py
def maxSubarraySumCircular(nums):
    global_max = nums[0]
    curr_max = -999999

    curr_min = 999999
    global_min = nums[0]

    total = 0

    for i in range(len(nums)):

        
        curr_min = min(curr_min + nums[i], nums[i])
        curr_max = max(curr_max + nums[i], nums[i])

        total += nums[i]

        global_max = max(global_max, curr_max)
        global_min = min(global_min, curr_min)

    return max(global_max, total-global_min) if global_max > 0 else global_max
    
nums = [-2,4,-5,4,-5,9,4]
print(maxSubarraySumCircular(nums))